ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,445 @@
+  /**
+   * Look for a profile from a given fiscal code
+   * @param fiscalCode a fiscal code identifying the user
+   * @returns either a user profile, a query error or a user-not-found error
+   */
+  const taskifiedFindProfile = (
+    fiscalCode: FiscalCode
+  ): TaskEither<
+    ActivityResultUserNotFound | ActivityResultQueryFailure,
+    Profile
+  > =>
+    fromQueryEither(
+      () => profileModel.findOneProfileByFiscalCode(fiscalCode),
+      "findOneProfileByFiscalCode"
+    ).foldTaskEither<
+      ActivityResultUserNotFound | ActivityResultQueryFailure,
+      Profile
+    >(
+      failure => fromEither(left(failure)),
+      maybeProfile =>
+        fromEither<ActivityResultUserNotFound, Profile>(
+          fromOption(
+            ActivityResultUserNotFound.encode({
+              kind: "USER_NOT_FOUND_FAILURE"
+            })
+          )(maybeProfile)
+        )
+    );
+
+  /**
+   * Retrieves all contents for provided messages
+   * @param messages
+   */
+  const getAllMessageContents = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<MessageContentWithId>
+  > =>
+    array.sequence(taskEither)(
+      messages.map(({ id: messageId }) =>
+        fromQueryEither(
+          () => messageModel.getContentFromBlob(blobService, messageId),
+          "messageModel.getContentFromBlob"
+        ).foldTaskEither<ActivityResultQueryFailure, MessageContentWithId>(
+          failure => fromEither(left(failure)),
+          maybeContent =>
+            fromEither(
+              fromOption(
+                ActivityResultQueryFailure.encode({
+                  kind: "QUERY_FAILURE",
+                  query: "messageModel.getContentFromBlob",
+                  reason: `Cannot find content for message ${messageId}`
+                })
+              )(maybeContent).map<MessageContentWithId>(
+                (content: MessageContent) => ({
+                  content,
+                  messageId
+                })
+              )
+            )
+        )
+      )
+    );
+
+  /**
+   * Given a list of messages, it queires for relative notifications
+   * @param messages
+   */
+  const findNotificationsForAllMessages = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<RetrievedNotification>
+  > =>
+    array
+      .sequence(taskEither)(
+        messages.map(m =>
+          fromQueryEither<ReadonlyArray<RetrievedNotification>>(
+            () =>
+              iteratorToArray(
+                notificationModel.findNotificationsForMessage(m.id)
+              ),
+            "findNotificationsForRecipient"
+          )
+        )
+      )
+      .foldTaskEither(
+        e => fromEither(left(e)),
+        arrayOfArray =>
+          // tslint:disable-next-line: readonly-array
+          fromEither(right(flatten(arrayOfArray as RetrievedNotification[][])))
+      );
```

Comment:
since you are converting back to taskEither, maybe I'm wrong but it looks like you can just use `map(flatten)` here

Reply:
you're right, `foldTaskEither` come from a version of the code which has been refactored now. No more reason for that, thanks